enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -296,9 +290,41 @@ able successfully reload the certificate on the route.
   > The ingress-to-route behaviour will remain as is i.e. it will not make use of
   > the newly introduced field.
 
-- What should be the behaviour of `ValidationHostUpdate()` when using `externalCertificate`?
-  > Addressed by introducing `ValidationHostExternalCertificate()` and which will execute
-  > prior to the `ValidationHostUpdate()` function.
+### Permission requirements on `routes/custom-host`
+- `create` is required to set `.spec.tls.externalCertificate` on route creation. 
+  [AllocateHost()](https://github.com/openshift/openshift-apiserver/blob/bd2a35e58172010c658f4d8f4dff8f9f0eac187d/pkg/route/apiserver/registry/route/strategy.go#L75) enforces this check.
+- Either `create` or `update` is required on route update with non-empty `.spec.tls.externalCertificate`. 
+  [ValidateHostUpdate](https://github.com/openshift/openshift-apiserver/blob/bd2a35e58172010c658f4d8f4dff8f9f0eac187d/pkg/route/apiserver/registry/route/strategy.go#L96) enforces this check.
+
+  - Few examples:
+
+    - Adding `.spec.tls.externalCertificate` to a route when no certificate was present.
+
+    - Adding `.spec.tls.externalCertificate` to a route when TLS configuration was nil.
+
+    - Switching from `.spec.tls.certificate/key` to `.spec.tls.externalCertificate`.
+
+    - Changing the secret name referenced by `.spec.tls.externalCertificate`.
+  
+    - **Even keeping the secret name unchanged, requires permission check. Since the state of the external secret cannot be verified.**
```

Comment:
This restriction really makes me nervous, for two reasons:

- This restriction potentially blocks _all_ updates to a route that specifies `spec.tls.externalCertificate` if the user does not have the "custom-host" permission.  Even updating an unrelated field or adding an annotation could be blocked.  However, we already discussed this concern here: https://github.com/openshift/openshift-apiserver/pull/407#discussion_r1571902991.  I'll continue to defer to the API reviewers if this is all right with them.  

- It isn't clear how this restriction on updates to the route improves security given that the user could update the secret without updating the route.  It seems that giving a user permission to _create_ a route with `spec.tls.externalCertificate` implicitly gives the user permission to _update_ the certificate, so maybe we need to require the user to have _update_ permission in order to specify `spec.tls.externalCertificate` at all.  

Can you make sure the EP explicitly calls out the first point (blocking all updates) and clearly addresses the second (how this restriction improves security)?

Reply:
> Can you make sure the EP explicitly calls out the first point (blocking all updates)

Sure, I'll add that information in the EP body. 

---

On the second concern, I think we might need to retain certain points from the `Exception in Validations between API server and the router` section which explains that "custom-host" permission check is intentionally done only on the API server and not the router as well. 

```
Post creation of the route if the permissions for `custom-host`
are revoked and the user edits the contents of the secret, the router would still be
able to successfully reload the certificate on the route.
```

> so maybe we need to require the user to have update permission in order to specify spec.tls.externalCertificate at all.

If a user has "custom-host" "create" permission, then "update" permission is optional to update the TLS certificate. Updating TLS certificate requires either `create` or `update` permission on the `routes/custom-host` sub-resource. So, users having "create" permission can do both create and update route with non-empty `spec.tls.externalCertificate`

Reply:
> > Can you make sure the EP explicitly calls out the first point (blocking all updates)
> 
> Sure, I'll add that information in the EP body.

Thanks!

> On the second concern, I think we might need to retain certain points from the `Exception in Validations between API server and the router` section which explains that "custom-host" permission check is intentionally done only on the API server and not the router as well.
> 
> ```
> Post creation of the route if the permissions for `custom-host`
> are revoked and the user edits the contents of the secret, the router would still be
> able to successfully reload the certificate on the route.
> ```
> 
> > so maybe we need to require the user to have update permission in order to specify spec.tls.externalCertificate at all.
> 
> If a user has "custom-host" "create" permission, then "update" permission is optional to update the TLS certificate. Updating TLS certificate requires either `create` or `update` permission on the `routes/custom-host` sub-resource. So, users having "create" permission can do both create and update route with non-empty `spec.tls.externalCertificate`


I see—create permission is already sufficient to update `spec.tls.certificate`, so it would be inconsistent to require update permission to specify an external certificate.  Moreover, requiring update permission to specify an external certificate doesn't prevent the user from updating the external certificate after the permission has been revoked.  Do I understand correctly?  

That leaves me with two unanswered questions:

- How does restricting updates on the _route_ improve security when someone could update the _secret_ for the external certificate?
- Can we prevent the route owner from updating the secret after custom-host create/update permissions have been revoked?  Or, do we just need to accept (and document!) that giving someone permission to use an external certificate effectively gives them irrevocable permission to update the certificate later on?

Reply:
> Sure, I'll add that information in the EP body.

Added

>Moreover, requiring update permission to specify an external certificate doesn't prevent the user from updating the external certificate after the permission has been revoked. Do I understand correctly?

Yes, that's correct. Users can update the external certificate (secret) even after the "custom-host" permission is revoked. 

> How does restricting updates on the route improve security when someone could update the secret for the external certificate?

You're right, it doesn't provide additional security in this context. As mentioned in the EP, this is a known limitation.

> Can we prevent the route owner from updating the secret after custom-host create/update permissions have been revoked? Or, do we just need to accept (and document!) that giving someone permission to use an external certificate effectively gives them irrevocable permission to update the certificate later on?

- One external certificate (secret) could be used (shared) by multiple routes. After the "custom-host" permissions are revoked for a certain route owner, blocking the secret update could cause issues for other route owners.

I haven't been able to identify an easier way to address this currently; it likely requires additional or potentially complex changes. This is an existing drawback, and since we haven't found a clear workaround, it should be documented.

Reply:
As discussed in the meeting, since this is an existing design choice and the bug is focused on making the `custom-host` permission consistent, does it make sense to discuss the resolution as part of a separate ticket if that is really a security concern?

Reply:
>How does restricting updates on the route improve security when someone could update the secret for the external certificate?
>
>Can we prevent the route owner from updating the secret after custom-host create/update permissions have been revoked? Or, do we just need to accept (and document!) that giving someone permission to use an external certificate effectively gives them irrevocable permission to update the certificate later on?

The following is the summary of our discussion with @deads2k over [Slack](https://redhat-internal.slack.com/archives/C06EK9ZH3Q8/p1728374123449109)

- Editor having edit access to Secrets within a namespace inherently gets a level of control over workloads in that namespace. They could also potentially change a Deployment's pod template to use a different image.
- Given the above, it's ultimately the responsibility of the cluster admin to manage and secure Secrets within a namespace appropriately, especially after revoking certain permissions. 
- Moreover, any namespace editor in a namespace that contains a custom host route has the power to change serving certificates for that custom route.

This means we don't need to prevent the route owner from updating the secret after custom-host create/update permissions have been revoked. The responsibility for managing access control ultimately lies with the cluster admin in this case.

Reply:
That summary makes sense, and it addresses my second question in https://github.com/openshift/enhancements/pull/1652#discussion_r1785305016 (whether we can prevent updates to the secret).  

Regarding my first question in https://github.com/openshift/enhancements/pull/1652#discussion_r1785305016 (how restricting updates on the route improve security), I think the answer is that it doesn't, right?  That is, restricting updates to routes with `externalCertificate` doesn't actually improve security.  

I get confused because the "Implementation Details" and "Permission requirements" sections say that the `custom-host` permission is required for route updates because the user could update the certificate in the secret after creating the route, but then the "Drawbacks" section explains that requiring this permission doesn't actually prevent the user from updating the certificate.  

However, if the API team is fine with this and I'm the only one who finds this confusing, then I won't hold this PR up any longer.

Reply:
I believe https://github.com/openshift/enhancements/pull/1652#discussion_r1791396123 was in regard to my first question about how the restriction improves security, so we can proceed with this EP for the purposes of fixing the associated bug report.